alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs

```diff
@@ -0,0 +1,466 @@
+import Shelley.Spec.Ledger.TxBody (unWdrl)
+import Shelley.Spec.Ledger.UTxO
+  ( UTxO (..),
+    balance,
+    txins,
+    txouts,
+    unUTxO,
+  )
+
+-- ============================================
+-- The uninhabited type that marks the STS rule
+
+data AlonzoUTXO era
+
+-- ==========================================================
+
+data UtxoPredicateFailure era
+  = BadInputsUTxO
+      !(Set (TxIn (Crypto era))) -- The bad transaction inputs
+  | OutsideValidityIntervalUTxO
+      !ValidityInterval -- transaction's validity interval
+      !SlotNo -- current slot
+  | MaxTxSizeUTxO
+      !Integer -- the actual transaction size
+      !Integer -- the max transaction size
+  | InputSetEmptyUTxO
+  | FeeTooSmallUTxO
+      !Coin -- the minimum fee for this transaction
+      !Coin -- the fee supplied in this transaction
+  | ValueNotConservedUTxO
+      !(Core.Value era) -- the Coin consumed by this transaction
+      !(Core.Value era) -- the Coin produced by this transaction
+  | WrongNetwork
+      !Network -- the expected network id
+      !(Set (Addr (Crypto era))) -- the set of addresses with incorrect network IDs
+  | WrongNetworkWithdrawal
+      !Network -- the expected network id
+      !(Set (RewardAcnt (Crypto era))) -- the set of reward addresses with incorrect network IDs
+  | OutputTooSmallUTxO
+      ![Core.TxOut era] -- list of supplied transaction outputs that are too small
+  | UpdateFailure (PredicateFailure (Core.EraRule "PPUP" era)) -- Subtransition Failures
+  | OutputBootAddrAttrsTooBig
+      ![Core.TxOut era] -- list of supplied bad transaction outputs
+  | TriesToForgeADA
+  | OutputTooBigUTxO
+      ![Core.TxOut era] -- list of supplied bad transaction outputs
+  | FeeNotBalancedUTxO
+      !Coin -- balance computed
+      !Coin -- the fee supplied in this transaction
+  | ScriptsNotPaidUTxO
+      !(UTxO era) -- The UTxO entries which have the wrong kind of script
+  | ExUnitsTooSmallUTxO
+      !ExUnits -- Max EXUnits from the protocol parameters
+      !ExUnits -- EXUnits supplied
+  deriving (Generic)
+
+deriving stock instance
+  ( Shelley.TransUTxOState Show era,
+    Show (PredicateFailure (Core.EraRule "PPUP" era))
+  ) =>
+  Show (UtxoPredicateFailure era)
+
+deriving stock instance
+  ( Shelley.TransUTxOState Eq era,
+    TransValue Eq era,
+    Eq (PredicateFailure (Core.EraRule "PPUP" era))
+  ) =>
+  Eq (UtxoPredicateFailure era)
+
+instance
+  ( Shelley.TransUTxOState NoThunks era,
+    NoThunks (PredicateFailure (Core.EraRule "PPUP" era))
+  ) =>
+  NoThunks (UtxoPredicateFailure era)
+
+-- ====================================
+
+minfee ::
+  ( HasField "_minfeeA" (Core.PParams era) Natural,
+    HasField "_minfeeB" (Core.PParams era) Natural,
+    HasField "_prices" (Core.PParams era) Prices
+  ) =>
+  Core.PParams era ->
+  Tx era ->
+  Coin
+minfee pp tx =
+  ((txsize tx) <×> a)
+    <+> b
+    <+> (scriptfee (getField @"_prices" pp) (txExunits (txbody tx)))
```

Comment:
the minfee calc is changed in my PR - to (I hope) what is correct, so this will have to be changed too